### PR TITLE
Inscryption: Implement new game

### PR DESCRIPTION
**Pull request by DrBibop**
## What is this fixing or adding?
Adds new world for the game Inscryption.
https://store.steampowered.com/app/1092790/Inscryption/

Learn more about the implementation here: https://github.com/Glowbuzz/Archipelago/blob/Inscryption/worlds/inscryption/docs/en_Inscryption.md

## How was this tested?
Has been available publicly and regularly tested by the community and the world developers since september of 2023 through our future-game-design thread. It has been included in multiple community asyncs as a stable unsupported game. Unit tests have also been added for logic.

### CODE REVIEW

**Comment by DrBibop** [comment]
Some of the latest changes do still need some testing but I'm starting the PR anyway in case anything needs changing

File: worlds/inscryption/Options.py
Comment (ScipioWright) [inline worlds/inscryption/Options.py:9]: Any specific reason you're not just importing DeathLink from Options?

File: worlds/inscryption/Options.py
Comment (ScipioWright) [inline worlds/inscryption/Options.py:35]:
```suggestion
    """Randomize codes and passwords in the game (clocks, safes, etc.)"""
```

File: worlds/inscryption/Rules.py
Comment (ScipioWright) [inline worlds/inscryption/Rules.py:114]:
```suggestion
        if self.world.options.epitaph_pieces_randomization == "all_pieces":
            return state.has("Epitaph Piece", self.player, 9)
        elif self.world.options.epitaph_pieces_randomization == "in_groups":
            return state.has("Epitaph Pieces", self.player, 3)
        else:
            return state.has("Epitaph Pieces", self.player, 1)
```

Would recommend using the string values for the option, or importing the option's class and doing something like
`if self.world.options.epitaph_pieces_randomization == EpitaphPiecesRandomization.option_all_pieces:`

File: worlds/inscryption/Rules.py
Comment (ScipioWright) [inline worlds/inscryption/Rules.py:132]: Very minor optimization, since `has_monocle` checks for one item instead of varied requirements.
```suggestion
    def has_tower_requirements(self, state: CollectionState) -> bool:
        return self.has_monocle(state) and self.has_act2_bridge_requirements(state)
```

File: worlds/inscryption/Rules.py
Comment (ScipioWright) [inline worlds/inscryption/Rules.py:156]:
```suggestion
    def has_pelts(self, state: CollectionState, count: int) -> bool:
        return state.has("Holo Pelt", self.player, count) and self.has_gems_and_battery(state)
```
Having 5 separate has_num_pelt helpers feels excessive -- would recommend condensing them into one function.

File: worlds/inscryption/__init__.py
Comment (ScipioWright) [inline worlds/inscryption/__init__.py:6]:
```suggestion
import .Rules
```
Not 100% sure this is correct since I'm not in front of an IDE atm

**Review by ScipioWright — COMMENTED** [review]
Out of time to keep reviewing for now, may review more later

File: worlds/inscryption/Options.py
Comment (DrBibop) [reply worlds/inscryption/Options.py:9]: Didn't know that was a thing. I'll look into it.

File: worlds/inscryption/Rules.py
Comment (DrBibop) [reply worlds/inscryption/Rules.py:156]: I agree that it's excessive. If I remember correctly, I think it was because my rules list is of the type dict[str, (CollectionState) -> bool] which doesn't allow me to add parameters like the number of pelts. If you know of a way, that would be great.

Comment (ScipioWright) [reply worlds/inscryption/Rules.py:156]: Ah, well, I have no idea then -- this is probably fine then to keep using that kind of structure. I'm definitely no stranger to having to do some hacky thing to keep everything else smooth

File: worlds/inscryption/Options.py
Comment (DrBibop) [reply worlds/inscryption/Options.py]: Changed to DeathLinkMixin. Gonna leave this as unresolved in case I didn't do it the right way.

File: worlds/inscryption/Options.py
Comment (ScipioWright) [inline worlds/inscryption/Options.py:107]:
```suggestion
from Options import DeathLink
...

class InscryptionOptions(DeathLinkMixin, PerGameCommonOptions):
    death_link: DeathLink
    act1_death_link_behaviour: Act1DeathLinkBehaviour
```

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py]: This was already fixed in my latest commit, no?

File: worlds/inscryption/Options.py
Comment (DrBibop) [reply worlds/inscryption/Options.py:107]: Doesn't DeathLinkMixin already add death_link to this list automatically? I used The Messenger as reference: https://github.com/ArchipelagoMW/Archipelago/blob/1e3a4b6db5acff8e82bdecdc79773f2923de919a/worlds/messenger/options.py#L221

File: worlds/inscryption/__init__.py
Comment (ScipioWright) [reply worlds/inscryption/__init__.py]: Sorry, I had done some more review and forgot to hit send, so when I reviewed the one thing today it sent the old ones too

File: worlds/inscryption/Options.py
Comment (ScipioWright) [reply worlds/inscryption/Options.py:107]: I have no idea how DeathLinkMixin works, I just know you removed death link from your options so I'd assume that would break it?

File: worlds/inscryption/Rules.py
Comment (remyjette) [reply worlds/inscryption/Rules.py:157]: Have `has_pelts` return a lambda that captures the count

Then, instead, of `"Act 3 - Trader 1": self.has_one_pelt`` you would call the function to get a lambda to assign
`"Act 3 - Trader 1": self.has_pelts(1)`

File: worlds/inscryption/Options.py
Comment (DrBibop) [reply worlds/inscryption/Options.py:107]: Judging from other implementations and from how death_link appears as an attribute in the class without adding it myself, it seems to work as it is currently.

Comment (ScipioWright) [reply worlds/inscryption/Options.py:107]: Ah, my bad then

File: worlds/inscryption/Rules.py
Comment (DrBibop) [reply worlds/inscryption/Rules.py]: Pushed in the latest commit. Works wonderfully!

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:6]: I'm assuming this should be okay as it is? Gonna mark this as resolved if that's the case.

File: worlds/inscryption/docs/en_Inscryption.md
Comment (DrBibop) [reply worlds/inscryption/docs/en_Inscryption.md:1]: Added a section for it. Let me know if it looks good.

**Comment by Exempt-Medic** [comment]
This has conflicts btw, also make sure you update https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/CODEOWNERS as well

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:62]: I don't love the hard-coded numbers here, but like, it works 🤷

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:81]: Is this intended to work where if less filler items need to be added than the list of filler items has, it _always_ picks the exact same set?

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:63]: Do you need the two `.value` here or the two in `set_rules`?

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:105]: I don't think you need be checking for this, just exclude the locations?

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:108]: Why can this be `None`, that seems strange to me?

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:126]: You could use this instead (check that the values are of the same format as  before though)
```suggestion
        return self.options.as_dict(
            "death_link",
            "act1_death_link_behaviour",
            "goal",
            "randomize_codes",
            "randomize_deck",
            "randomize_sigils",
            "optional_death_card",
            "skip_tutorial",
            "skip_epilogue",
            "epitaph_pieces_randomization"
        )
```

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:94]: You can use the helper function instead:
```suggestion
            region = self.get_region(region_name)
```

File: worlds/inscryption/Options.py
Comment (Exempt-Medic) [inline worlds/inscryption/Options.py:10]: This is True by default with `Choice` options and can be removed from all Choice options here
```suggestion
```

File: worlds/inscryption/Options.py
Comment (Exempt-Medic) [inline worlds/inscryption/Options.py:63]: Out of curiosity, how does this work with DeathLinks only extinguishing _one_ candle? Does it have to extinguish the last one specifically?

File: worlds/inscryption/Options.py
Comment (Exempt-Medic) [inline worlds/inscryption/Options.py:104]: You correctly have two newlines everywhere else in the file
```suggestion


```

File: worlds/inscryption/Rules.py
Comment (Exempt-Medic) [inline worlds/inscryption/Rules.py:107]: This can use `has_all`
```suggestion
        return state.has_all(("Oil Painting's Clover Plant", "Squirrel Totem Head"), self.player)
```

File: worlds/inscryption/Rules.py
Comment (Exempt-Medic) [inline worlds/inscryption/Rules.py:118]: This can use `has_all`
```suggestion
        return state.has_all(("Camera Replica", "Pile Of Meat"), self.player)
```

**Comment by DrBibop** [comment]
> This has conflicts btw

Should I leave the README alone and let someone else add the game to the supported list after its merged?

File: worlds/inscryption/Rules.py
Comment (Exempt-Medic) [inline worlds/inscryption/Rules.py:115]: Two things. First, you could do this instead (also, I don't think you need `.value`):
```suggestion
        if self.world.options.epitaph_pieces_randomization == EpitaphPiecesRandomization.option_all_pieces:
            return state.has("Epitaph Piece", self.player, 9)
        if self.world.options.epitaph_pieces_randomization == EpitaphPiecesRandomization.option_in_groups:
            return state.has("Epitaph Pieces", self.player, 3)
        return state.has("Epitaph Pieces", self.player, 1)
```
Second, this is checking options inside of an access rule, which may get checked hundreds of time, so the if chain can be expensive, it would be best to refactor this so that you're not checking options inside the rule, either through lambdas or assigning a value to instances of your world class and using that.

File: worlds/inscryption/Rules.py
Comment (Exempt-Medic) [inline worlds/inscryption/Rules.py:188]: I'm again unsure why this can ever be `None` and isn't just:
```suggestion
    def set_painting_rules(self) -> None:
        set_rule(self.world.get_location("Act 1 - Painting 2"), self.has_useful_act1_items)
        set_rule(self.world.get_location("Act 1 - Painting 3"), self.has_useful_act1_items)
```

It seems like this is trying to prevent certain option combinations from failing, but if that's the case, you should simply be checking for those options and throwing an OptionError in `generate_early` instead of disregarding the options a player used

**Comment by Exempt-Medic** [comment]
> > This has conflicts btw
> 
> Should I leave the README alone and let someone else add the game to the supported list after its merged?

Updating it yourself is best

**Comment by DrBibop** [comment]
Anyway, resolved the conflict

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
Some small comments

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:40]: Perhaps, but this was copied directly from the Steam store page. Should we still add it?

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:62]: wdym by hard-coded numbers? Do you have an alternate solution?

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:81]: The only situation where that happens is when the goal is Act 1 only, which would only have room for 1 filler item. Are you saying we should be picking randomly from the list of fillers?

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:63]: Oh, nice catch. Not sure why I was using it in some places and not in others.

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:40]: Oh LOL, in that case, yeah, it's totally fine

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:62]: The lines like `useful_items[len(act1_items) + 2]` where it's taking advantage of the very particular order/structure of the items instead of searching by name. It's just easier to mess up, especially if you ever change anything in the tables. I don't think it _needs_ to change, but figured I should mention it

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:81]: I'm just saying it seems like a strange decision to me. If you're basing the fact that this works off of how your *current* options work, then you might as well just add filler_count // 2 of one filler and filler_count // 2 of the other (and then add one as needed)

File: worlds/inscryption/test/TestGoal.py
Comment (DrBibop) [reply worlds/inscryption/test/TestGoal.py:10]: This basically adds every required item one by one and tests that the game is not beatable before collecting it. After everything is collected, it tests that the game is now beatable. The epitaph pieces are separate from the "required_items_all_acts" simply because the item name depends on the option as you can see in the other tests.

Comment (Exempt-Medic) [reply worlds/inscryption/test/TestGoal.py:10]: I think these should be using AssertAccessDependency then, since you're really only testing the _entire_ set and not each item

**Comment by Exempt-Medic** [comment]
Two things I think you may want to consider adding:
1. `get_filler_item_name`
This makes it so that when a filler item is required from your world (like with itemlinks, plando, or a few other things) it will only pull the items you want it to instead of any random item, including progression items.
2. `start_inventory_from_pool`
This is an opt-in option, but it's very easy to setup ( for an example: https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/tunic/options.py#L194 ). This would allow players to have a `start_inventory` option that actually removes items from the pool instead of just adding duplicates (this would also benefit from `get_filler_item_name` since it uses that). For another simple example from TUNIC: https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/tunic/__init__.py#L320

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:81]: Yeah which is why I brought up my initial comment that the last set isn't picked randomly, but if you don't see that as a problem, I'd understand that. Just wanted to make sure you were aware / it was intended

File: worlds/inscryption/Options.py
Comment (DrBibop) [reply worlds/inscryption/Options.py:63]: Yes, that'd be how it works.

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:62]: To be fair, this stems from my lack of Python experience. What would be the proper way to search by name, then? It's a list of ItemDict which contains a name, count and classification.

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:105]: What check specifically?

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:126]: Not all options are to be added in the slot data, so I don't think this can be applied.

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:62]: I'll be honest, I was trying to think of better ways, but with the way it's structured, I honestly think other ways might just be worse. So how it is seems fine, and it's mostly personal preference, AP has a good chunk of magic numbers in its worlds and this one is honestly quite tame.

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:126]: This would only add the ones I listed

File: worlds/inscryption/test/TestGoal.py
Comment (DrBibop) [reply worlds/inscryption/test/TestGoal.py:10]: Should I just remove the False checks while keeping the final one? Or are these tests just not needed?

Comment (Exempt-Medic) [reply worlds/inscryption/test/TestGoal.py:10]: I think a better way to do these would be to give yourself every item, assert beatability, then remove one, assert _not_ beatability, put it back and remove the next (different) one.

File: worlds/inscryption/__init__.py
Comment (DrBibop) [reply worlds/inscryption/__init__.py:126]: Ah, my mistake then.

File: worlds/inscryption/test/TestGoal.py
Comment (DrBibop) [reply worlds/inscryption/test/TestGoal.py:10]: Smart. Gonna give it a try.

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [reply worlds/inscryption/__init__.py:105]: Might be easier to explain this whole thing in the server since it's a few different things at once for me

File: worlds/inscryption/__init__.py
Comment (Exempt-Medic) [inline worlds/inscryption/__init__.py:64]: Do you need the outer if here?
```suggestion
        if self.options.epitaph_pieces_randomization == EpitaphPiecesRandomization.option_all_pieces:
            self.required_epitaph_pieces_name = "Epitaph Piece"
            self.required_epitaph_pieces_count = 9
        elif self.options.epitaph_pieces_randomization == EpitaphPiecesRandomization.option_in_groups:
            self.required_epitaph_pieces_name = "Epitaph Pieces"
            self.required_epitaph_pieces_count = 3
        else:
            self.required_epitaph_pieces_name = "Epitaph Pieces"
            self.required_epitaph_pieces_count = 1
```

**Comment by Exempt-Medic** [comment]
Another thing I don't know if you'd considered, but you could add location and item groups, things like "Act 2" locations and "Card" items and other things like that

**Review by Exempt-Medic — APPROVED** [review]
Reviewed all of the code. I do not know French and so did not review the French setup guide. Merged into `main` and performed 1000 random single-player generations, 100 10-player generations, and 50 100-player generations (modified to error if a `full` player has inaccessible locations). An error was hit quite early and is fixed in [my PR](https://github.com/Glowbuzz/Archipelago/pull/1) that definitely does more than things than it should. With the issue found there fixed, no other errors were encountered during generations. All of my previous comments have been addressed other than potentially adding in item and location groups

**Comment by DrBibop** [comment]
If everything looks okay, I'm planning on doing another beta release with the recent changes and undraft this PR after

File: worlds/inscryption/Options.py
Comment (NewSoupVi) [inline worlds/inscryption/Options.py:102]: This is not a thing other core maintainers usually insist on & moreso my personal opinion, but I'm the core maintainer you got, sooo ;)

Imo, Choice options should always explicitly define `default`. (& basically any option other than Toggle and DefaultOnToggle)

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
Literally one incredibly minor comment. Other than that, this is super good to go

File: worlds/inscryption/Options.py
Comment (DrBibop) [reply worlds/inscryption/Options.py:106]: Done 👍

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**
